apps/prairielearn/src/ee/lib/terms.ts

```diff
@@ -0,0 +1,44 @@
+import { type Response } from 'express';
+import { callRow } from '@prairielearn/postgres';
+
+import { ModeSchema, type User } from '../../lib/db-types';
+
+export function hasUserAcceptedTerms(user: User): boolean {
+  // At the moment, we only have one revision of our terms and conditions, so
+  // all we care about is whether the user has accepted the terms at any point.
+  // In the future, if we add revisions, we could change this to check for
+  // acceptance after the date that the most recent revision went into effect.
+  return user.terms_accepted_at !== null;
+}
+
+/**
+ * Determines whether the user should be redirected to the terms acceptance page.
+ * This is the case if the user has not yet accepted the terms and the user is
+ * accessing the site in Public mode. We want to avoid prompting for terms
+ * acceptance if the user is accessing the site in Exam mode, since they may not
+ * have time to read the terms or accept them before the exam starts, or the network
+ * may not even allow them to the terms pages.
+ *
+ * @param user The user to check
+ * @param ip The IP address of the request
+ * @returns Whether the user should be redirected to the terms acceptance page
+ */
+export async function shouldRedirectToTermsPage(user: User, ip: string) {
+  if (user.terms_accepted_at !== null) return false;
+
+  const mode = await callRow('ip_to_mode', [ip, new Date(), user.user_id], ModeSchema);
```

Comment:
I'm doubting if this will actually do exactly what we want. `ip_to_mode` only considers "active" reservations, that is, ones with a defined `access_start` and `access_end`. So, this doesn't catch the case of a student signing in and/or authing to PT through PL before a proctor has started the exam. I'm wondering if this behavior has implications outside of this PR. That is: if a student in a testing center directly accesses PL before their exam has started, would they be in `Public` mode and thus able to access arbitrary course content?

Reply:
See https://github.com/PrairieLearnInc/PrairieTest/issues/802 -- let's take the discussion there?

Reply:
Fixed in #9590